From 0a6aba1014cf0a83acb5da155c6f949ce1daaf54 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Fri, 26 Jun 2020 11:51:27 +0300
Subject: [PATCH] [SW-14252] preboot: goya2: handle multiple images start

As boot FIT contains multiple images which are copied into respective
load address, this patch adds preboot_fit_start_images function to
handle multiple boot FIT images. preboot_fit_start_images function of
goya2_arc_gnu HL_TARGET is also modified to boot ARC0 with zephyr pid
image and jump ARC1 to zephyr mgmt image along with booting UBOOT on
ARM.

Change-Id: I229559b60ec22d9382080972da96e9e5e9918562
---

diff --git a/gaudi/common.c b/gaudi/common.c
index af1cbb8..6f0f564 100644
--- a/gaudi/common.c
+++ b/gaudi/common.c
@@ -22,6 +22,7 @@
 #include <uart.h>
 #include <string.h>
 #include <ppboot/spi.h>
+#include <preboot/fit.h>
 #include <thermal.h>
 
 #include "common.h"
@@ -249,6 +250,26 @@
 	image_entry();
 }
 
+void preboot_fit_start_images(struct fit_image *image)
+{
+	void (*image_entry)(void) = NULL;
+	u32 i;
+
+	for (i = 0; i < FIT_MAX_IMAGES; i++) {
+		if (strcmp(image[i].arch, "arm64"))
+			continue;
+
+		image_entry = (void *)image[i].loadaddr;
+		break;
+	}
+
+	/*
+	 * After image_entry, control is not expected to return
+	 * here and continue.
+	 */
+	image_entry();
+}
+
 void fit_copy_image(u64 dst, const void *buf, int len)
 {
 	preboot_puts_hex_raw("FIT loadaddr: ", (u32)(dst >> 32), false);
diff --git a/goya/common.c b/goya/common.c
index 5e9c525..ecc8d62 100644
--- a/goya/common.c
+++ b/goya/common.c
@@ -19,6 +19,7 @@
 #include <uart.h>
 #include <string.h>
 #include <ppboot/spi.h>
+#include <preboot/fit.h>
 #include <thermal.h>
 #include <string.h>
 
@@ -193,6 +194,26 @@
 	image_entry();
 }
 
+void preboot_fit_start_images(struct fit_image *image)
+{
+	void (*image_entry)(void) = NULL;
+	u32 i;
+
+	for (i = 0; i < FIT_MAX_IMAGES; i++) {
+		if (strcmp(image[i].arch, "arm64"))
+			continue;
+
+		image_entry = (void *)image[i].loadaddr;
+		break;
+	}
+
+	/*
+	 * After image_entry, control is not expected to return
+	 * here and continue.
+	 */
+	image_entry();
+}
+
 void fit_copy_image(u64 dst, const void *buf, int len)
 {
 	preboot_puts_hex_raw("FIT loadaddr: ", (u32)(dst >> 32), false);
diff --git a/goya2/common.c b/goya2/common.c
index def71ec..ce194ba 100644
--- a/goya2/common.c
+++ b/goya2/common.c
@@ -24,6 +24,7 @@
 #include <uart.h>
 #include <string.h>
 #include <ppboot/spi.h>
+#include <preboot/fit.h>
 
 #include "common.h"
 
@@ -256,6 +257,26 @@
 	image_entry();
 }
 
+void preboot_fit_start_images(struct fit_image *image)
+{
+	void (*image_entry)(void) = NULL;
+	u32 i;
+
+	for (i = 0; i < FIT_MAX_IMAGES; i++) {
+		if (strcmp(image[i].arch, "arm64"))
+			continue;
+
+		image_entry = (void *)image[i].loadaddr;
+		break;
+	}
+
+	/*
+	 * After image_entry, control is not expected to return
+	 * here and continue.
+	 */
+	image_entry();
+}
+
 void fit_copy_image(u64 dst, const void *buf, int len)
 {
 	preboot_puts_hex_raw("FIT loadaddr: ", (u32)(dst >> 32), false);
diff --git a/goya2_arc_gnu/common.c b/goya2_arc_gnu/common.c
index ad197d4..b119690 100644
--- a/goya2_arc_gnu/common.c
+++ b/goya2_arc_gnu/common.c
@@ -9,6 +9,8 @@
 #include <psoc_global_conf_regs.h>
 #include <psoc_gpio1_regs.h>
 #include <psoc_qspi_regs.h>
+#include <psoc_arc0_cfg_regs.h>
+#include <psoc_arc0_cfg_masks.h>
 #include <pcie_aux_regs.h>
 #include <pcie_wrap_regs.h>
 #include <pcie_core_regs.h>
@@ -27,7 +29,7 @@
 #include <pll.h>
 #include <uart.h>
 #include <string.h>
-
+#include <preboot/fit.h>
 #include "common.h"
 
 #define FW_WORD_NUM		0x2200	/* FW_SIZE / 2 */
@@ -133,21 +135,46 @@
 	memcpy((void *)arc_dst, buf, len);
 }
 
-void image_start(u64 loadaddr)
+void preboot_fit_start_images(struct fit_image *image)
 {
-	u32 lsb = (u32)loadaddr, msb = (u32)(loadaddr >> 32);
+	void (*image_entry)(void);
+	u32 lsb, msb;
+	u32 i;
 
-	/* Set CPU0 Reset address to U-Boot */
-	regmap_writel(lsb, mmCPU_CA53_CFG_RST_ADDR_LSB_0);
-	regmap_writel(msb, mmCPU_CA53_CFG_RST_ADDR_MSB_0);
+	for (i = 0; i < FIT_MAX_IMAGES; i++) {
+		if (strcmp(image[i].arch, "arm64"))
+			continue;
 
-	/* Wake up CPU0, after we put all CPUs in reset */
-	regmap_writel(0x0, mmCPU_CA53_CFG_ARM_RST_CONTROL);
-	regmap_writel(0x10111, mmCPU_CA53_CFG_ARM_RST_CONTROL);
+		lsb = (u32)image[i].loadaddr;
+		msb = (u32)(image[i].loadaddr >> 32);
+		/* Set CPU0 Reset address to U-Boot */
+		regmap_writel(lsb, mmCPU_CA53_CFG_RST_ADDR_LSB_0);
+		regmap_writel(msb, mmCPU_CA53_CFG_RST_ADDR_MSB_0);
 
-	/* Main Loop */
-	while (1)
+		/* Wake up CPU0, after we put all CPUs in reset */
+		regmap_writel(0x0, mmCPU_CA53_CFG_ARM_RST_CONTROL);
+		regmap_writel(0x10111, mmCPU_CA53_CFG_ARM_RST_CONTROL);
+		break;
+	}
+
+	/*
+	 * Set ARC0 in running state
+	 * 1. Set run bit in RUN_HALT_REQ register
+	 * 2. Wait for ACK from core in RUN_HALT_ACK register
+	 * 3. Clear run bit in RUN_HALT_REQ register
+	 */
+	regmap_writel(BIT(0), mmPSOC_ARC0_CFG_RUN_HALT_REQ);
+	while (!(regmap_readl(mmPSOC_ARC0_CFG_RUN_HALT_ACK) &
+				PSOC_ARC0_CFG_RUN_HALT_ACK_RUN_ACK_MASK))
 		;
+	regmap_writel(0x0, mmPSOC_ARC0_CFG_RUN_HALT_REQ);
+
+	/*
+	 * Jump to __reset of zephyr. After image_entry, control is not
+	 * expected to return here and continue.
+	 */
+	image_entry = (void *) (*(u32 *)ARC_ICCM_BASE);
+	image_entry();
 }
 
 void ppboot_cp_preboot(u64 loadaddr, u32 offset, u32 size, char *name)
diff --git a/goya2_arc_gnu/common.h b/goya2_arc_gnu/common.h
index bc04b5f..4315317 100644
--- a/goya2_arc_gnu/common.h
+++ b/goya2_arc_gnu/common.h
@@ -14,6 +14,7 @@
 #define PCIE_FW_SRAM_BASE	0x1BFF0000
 #define FW_SRAM_BASE		0x1B050000
 #define SP_SRAM_BASE		0x1BFE0000
+#define ARC_ICCM_BASE		0x1B020000
 #define SP_SRAM_PREBOOT_OFFSET	0x1000
 #define PPBOOT_TEXT_BASE	(FW_SRAM_BASE + 0xC)
 #define PREBOOT_TEXT_BASE	(SP_SRAM_BASE + SP_SRAM_PREBOOT_OFFSET + 0x8)
diff --git a/preboot/fit.h b/preboot/fit.h
index 4f39242..59eaef5 100644
--- a/preboot/fit.h
+++ b/preboot/fit.h
@@ -9,7 +9,6 @@
 #define _HL_FIT_H_
 
 #define FIT_MAX_IMAGES		8
-#define FIT_UBOOT_INDEX		0
 
 struct fit_image {
 	u64 loadaddr;
@@ -19,5 +18,6 @@
 int fit_load_images(const void *fit, struct fit_image *images);
 int fit_check_format(const void *fit);
 void fit_copy_image(u64 dst, const void *buf, int len);
+void preboot_fit_start_images(struct fit_image *image);
 
 #endif /* _HL_FIT_H_ */
diff --git a/preboot/module_lkd.c b/preboot/module_lkd.c
index ee6d087..b4e3836 100644
--- a/preboot/module_lkd.c
+++ b/preboot/module_lkd.c
@@ -97,7 +97,7 @@
 		wfe();
 		break;
 	case KMD_MSG_FIT_RDY:
-		image_start(sts.images[FIT_UBOOT_INDEX].loadaddr);
+		preboot_fit_start_images(sts.images);
 		/* no break */
 	default:
 		break;
